Type-annotate SkipMixin and BaseXCom

uranusjr opened the pull request:
Extracted from #19825, these are required to make Mypy pass checking the two files.

uranusjr committed:
Typing fixes for SkipMixin

uranusjr committed:
Type-annotate XCom get_*, set, delete, and clear

Plus some minor fixups to make the entire file pass Mypy.

ashb commented on airflow/models/skipmixin.py:
```diff
@@ -17,7 +17,9 @@
-from typing import TYPE_CHECKING, Iterable, Union
+from typing import TYPE_CHECKING, Iterable, Optional, Sequence, Union
+
+import pendulum
```
Since we already have `if TYPE_CHECKING` on L30 move this to that block?

uranusjr replied:
Moved

ashb commented on airflow/models/xcom.py:
```diff
@@ -99,29 +137,63 @@ def set(cls, key, value, task_id, dag_id, execution_date=None, run_id=None, sess
-
+        new = cast(Any, cls)(  # Work around Mypy complaining model not defining '__init__'.
+            key=key,
+            value=cls.serialize_value(value),
```
This is different -- before it was just doing `value=value`.

```suggestion
            value=value,
```

I think?

uranusjr replied:
Hmm not sure how this got here.

uranusjr replied:
Ah I see, wrong copy-paste. I merged line 100/138 above to this in the other PR, but forgot to delete that line when copying the new implementation.

ashb commented on airflow/models/xcom.py:
```diff
@@ -376,4 +505,7 @@ def resolve_xcom_backend():
 
 
-XCom = resolve_xcom_backend()
+if TYPE_CHECKING:
+    XCom = BaseXCom  # Hack to avoid Mypy "Variable 'XCom' is not valid as a type".
+else:
+    XCom = resolve_xcom_backend()
```
I think this might be causing docs to fail too, as it's somehow finding two places for all the methods on XCom.

(I'm guessing as to cause, doc build is failing though)

uranusjr replied:
I think this is fine, the cause is actually the overloads:

```
WARNING: duplicate object description of airflow.models.xcom.BaseXCom.clear, other instance in _api/airflow/models/xcom/index, use :noindex: for one of them

File path: apache-airflow/_api/airflow/models/xcom/index.rst (215)

 210 |    .. classmethod:: clear(cls, execution_date: pendulum.DateTime, dag_id: str, task_id: str, session: Optional[Session] = None)
 211 | 
 212 | 
 213 | 
 214 |    
>215 |    .. classmethod:: clear(cls, execution_date: Optional = None, dag_id: Optional = None, task_id: Optional = None, *, run_id: Optional = None, session: Session)
 216 | 
 217 |       Clears all XCom data from the database for the task instance
 218 | 
 219 |       ``run_id`` and ``execution_date`` are mutually exclusive.
```

I think generated by sphinx-autoapi. I’d expect it to support this _years_ old feature; it’s very disappoiting.

uranusjr replied:
Well actually it does support it since last year, but we’re pinning an extremely old version.

ashb replied:
Every so often I've tried upgrading sphinx-autoapi, but last _n_ times I tried there was a bug that meant it didn't build docs on a more recent version :(

uranusjr replied:
Why are we using autoapi in the first place? Do we have any non-Python APIs that can’t be documented with sphinx.ext.autodoc?

potiuk replied:
I think this is question mostly to @mik-laj who knows more of the context. But I share the same sentiment and experience as @ashb - any time I tried I backtracked so far. And yea - we are using extremely old version.

uranusjr committed:
Add AutoAPI monkeypatch to allow manually skipping

This patch makes Sphinx-AutoAPI to skip anything if its docstring
contains ':sphinx-autoapi-skip:' (no quotes, surrounded by whitespaces).

ashb commented on airflow/models/skipmixin.py:
```diff
@@ -66,9 +67,10 @@ def _set_state_to_skipped(self, dag_run: "DagRun", tasks: "Iterable[BaseOperator
+        execution_date: "DateTime",
+        tasks: Sequence["BaseOperator"],
+        *,
+        session: "Session",
```
We probably shouldn't make session a kw-only arg, as that _might_ break someone's custom operator :(

ashb commented on airflow/models/xcom.py:
```diff
@@ -79,144 +80,247 @@ def init_on_load(self):
+        execution_date: Optional[datetime.datetime] = None,
+        *,
+        run_id: Optional[str] = None,
+        session: Session,
```
Ditto

uranusjr committed:
Revert session kwarg changes
